### PR TITLE
resolve issue #60

### PR DESCRIPTION
**Pull request by nurzhan-saktaganov**
- DiscoveryTimeout by default is 1 minute (zero DiscoveryTimeout is not allowed)
- cronDiscovery: make it panic-tolerant
- bugfix: remove defer in for loop

### CODE REVIEW

File: discovery.go [hunk 1]
```diff
@@ -212,31 +212,48 @@ func (r *Router) DiscoveryAllBuckets(ctx context.Context) error {
...
+		// (Similar to pcall in lua/tarantool)
+		func() {
 			defer func() {
-				r.log().Infof(ctx, "discovery done since %s", time.Since(tStartDiscovery))
+				if recovered := recover(); recovered != nil {
+					// Another one panic may happen due to log function below (e.g. bug in log().Errorf), in this case we have two options:
+					// 1. recover again and log nothing: panic will be muted and lost
+					// 2. don't try to recover, we hope that the second panic will be logged somehow by go runtime
+					// So, we choose the second behavior
+					r.log().Errorf(ctx, "ACHTUNG: something unexpected has happened in cronDiscovery(%d): %v", iteration, recovered)
```
Comment (KaymeKaydex) [inline discovery.go:240]: I think we shouldn't use the expression ACHTUNG and replace it with PANIC

It’s also worth logging the stack trace

File: vshard.go [hunk 2]
```diff
@@ -232,6 +230,10 @@ func prepareCfg(cfg Config) (Config, error) {
 		return Config{}, fmt.Errorf("%v: %v", ErrInvalidConfig, err)
 	}
 
+	if cfg.DiscoveryTimeout == 0 {
+		cfg.DiscoveryTimeout = 1 * time.Minute
```
Comment (KaymeKaydex) [inline vshard.go:234]: it is worth transferring the default value ​​to a constant

File: discovery.go [hunk 3]
```diff
@@ -212,31 +212,48 @@ func (r *Router) DiscoveryAllBuckets(ctx context.Context) error {
...
+func (r *Router) cronDiscovery(ctx context.Context) {
+	var iteration uint64
+
 	for {
 		select {
 		case <-ctx.Done():
 			r.metrics().CronDiscoveryEvent(false, 0, "ctx-cancel")
-
-			return ctx.Err()
+			r.log().Infof(ctx, "cron discovery has been stopped after %d iterations", iteration)
```
Comment (KaymeKaydex) [inline discovery.go:223]: Add the [DISCOVERY] tag at the beginning of the discovery log

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,12 +10,14 @@ BUG FIXES:
...
 
 * Support new Sprintf-like logging interface (#48)
+* DiscoveryTimeout by default is 1 minute (zero DiscoveryTimeout is not allowed #60)
 
 REFACTOR:
 
 * resolve issue #38: simplify DiscoveryAllBuckets and remove suspicious if
 * resolve issue #46: drastically simplify RouterMapCallRWImpl and added tests with real tnt
 * Use typed nil pointers instead of memory allocation for EmptyMetrics and emptyLogger structs
+* rename stattCronDiscovery to cronDiscovery and make it panic-tolerant
```
Comment (KaymeKaydex) [inline CHANGELOG.md:20]:
```suggestion
* rename startCronDiscovery to cronDiscovery and make it panic-tolerant
```

File: discovery.go [hunk 5]
```diff
@@ -212,31 +212,48 @@ func (r *Router) DiscoveryAllBuckets(ctx context.Context) error {
 	return nil
 }
 
-// startCronDiscovery is discovery_service_f analog with goroutines instead fibers
-func (r *Router) startCronDiscovery(ctx context.Context) error {
+// cronDiscovery is discovery_service_f analog with goroutines instead fibers
+func (r *Router) cronDiscovery(ctx context.Context) {
+	var iteration uint64
```
Comment (KaymeKaydex) [inline discovery.go:217]: iterationCount

**Review by KaymeKaydex — CHANGES_REQUESTED** [review]

File: discovery.go (hunk 1 above)
Comment (nurzhan-saktaganov) [reply discovery.go]: Done

File: vshard.go (hunk 2 above)
Comment (nurzhan-saktaganov) [reply vshard.go]: Done

File: discovery.go (hunk 3 above)
Comment (nurzhan-saktaganov) [reply discovery.go]: Done

File: CHANGELOG.md (hunk 4 above)
Comment (nurzhan-saktaganov) [reply CHANGELOG.md]: Done

File: discovery.go (hunk 5 above)
Comment (nurzhan-saktaganov) [reply discovery.go]: Done

File: discovery.go [hunk 6]
```diff
@@ -212,31 +213,49 @@ func (r *Router) DiscoveryAllBuckets(ctx context.Context) error {
...
+		func() {
 			defer func() {
-				r.log().Infof(ctx, "discovery done since %s", time.Since(tStartDiscovery))
+				if recovered := recover(); recovered != nil {
+					// Another one panic may happen due to log function below (e.g. bug in log().Errorf), in this case we have two options:
+					// 1. recover again and log nothing: panic will be muted and lost
+					// 2. don't try to recover, we hope that the second panic will be logged somehow by go runtime
+					// So, we choose the second behavior
+					r.log().Errorf(ctx, "[DISCOVERY] something unexpected has happened in cronDiscovery(%d): panic %v", iterationCount, recovered)
+					r.log().Errorf(ctx, "[DISCOVERY] stacktrace for previous panic: %s", string(debug.Stack()))
```
Comment (KaymeKaydex) [inline discovery.go:242]: let's make it a single log

Comment (nurzhan-saktaganov) [reply discovery.go]: Done

File: CHANGELOG.md [hunk 7]
```diff
@@ -10,13 +10,15 @@ BUG FIXES:
 FEATURES:
 
 * Support new Sprintf-like logging interface (#48)
+* DiscoveryTimeout by default is 1 minute (zero DiscoveryTimeout is not allowed #60)
```
Comment (KaymeKaydex) [inline CHANGELOG.md:14]: add 1 more line about new discovery logs tag

**Review by KaymeKaydex — APPROVED** [review]

**Pull request merged**